tools/elfdeps.c

```diff
@@ -33,6 +38,109 @@ typedef struct elfInfo_s {
     ARGV_t provides;
 } elfInfo;
 
+/*
+ * If filename contains ".so" followed by a version number, return
+ * a copy of the version number.
+ */
+static char *getLibtoolVer(const char *filename)
+{
+    const char *so;
+    char dest[PATH_MAX];
+    int destsize = 0;
+    int found_digit = 0, found_dot = 0;
+
+    destsize = readlink(filename, dest, PATH_MAX);
+    if (destsize > 0) {
+	dest[destsize] = 0;
+	filename = dest;
+    }
+    /*
+     * Start from the end of the string.  Verify that it ends with
+     * numbers and dots, preceded by ".so.".
+     */
+    so = filename + strlen(filename);
+    while (so > filename+2) {
+	if (*so == '.') {
+	    found_dot++;
+	    so--;
+	    continue;
+	} else if (strchr("0123456789", *so)) {
+	    found_digit++;
+	    so--;
+	    continue;
+	} else if (strncmp(so-2, ".so.", 4) == 0) {
+	    so+=2;
+	    if (found_digit && found_dot > 1) {
+		return strdup(so);
+	    }
+	    break;
+	} else {
+	    break;
+	}
+    }
+    return NULL;
+}
+
+/*
+ * Rather than re-implement path searching for shared objects, use
+ * dlmopen().  This will still perform initialization and finalization
+ * functions, which isn't necessarily safe, so do that in a separate
+ * process.
+ */
+static char *getLibtoolVerFromShLink(const char *filename)
+{
+#if defined(HAVE_DLMOPEN) && defined(HAVE_DLINFO)
+    char dest[PATH_MAX];
+    int pipefd[2];
+    pid_t cpid;
+
+    if (pipe(pipefd) == -1) {
+	return NULL;  /* Should this be a fatal error instead? */
+    }
+    cpid = fork();
+    if (cpid == -1) {
+	return NULL;  /* Should this be a fatal error instead? */
+    }
+    if (cpid == 0) {
+	void *dl_handle;
+	struct link_map *linkmap;
+	char *version = NULL;
+
+	close(pipefd[0]);
+	dl_handle = dlmopen(LM_ID_NEWLM, filename, RTLD_LAZY);
+	if (dl_handle == NULL) _exit(0);
+	if (dlinfo(dl_handle, RTLD_DI_LINKMAP, &linkmap) != -1) {
+	    version = getLibtoolVer(linkmap->l_name);
+	}
+	if (version)
+	    (void) write(pipefd[1], version, strlen(version));
+	close(pipefd[1]);
+	free(version);
+	dlclose(dl_handle);
+	_exit(0);
+    } else {
+	ssize_t len;
+	close(pipefd[1]);
+	dest[0] = 0;
+	len = read(pipefd[0], dest, sizeof(dest));
+	if (len > 0) dest[len] = 0;
+	close(pipefd[0]);
+	wait(NULL);
```

Comment:
Sorry I forgot to mention this yesterday, but we'll want to check the exit status of the child here. The segfault due to missing version was getting silently unnoticed because of that, I only noticed because execution was taking *far* too long for such a simple thing, and investigation turned up that segfault.

Reply:
I'm curious about this, because if the child process segfaults the parent's read() should return with no data read, and the process should continue without the version number.  It sounds like you're saying that it blocked on that read instead, and I'm not sure why that would happen.

Reply:
Oh, it was falling through the read() alright, the delay comes from systemd-coredump kicking in on the child, and then continuing as if nothing had happened. Talking about 0.01s vs 0.5s execution time here, which is enough to be visually noticeable, very much so if you happen to feed more than one file at a time. 

The point though, the child crashing for any reason at all is something we want to notice and fail with an error in the parent.